### PR TITLE
LEST-29 - Number matchers

### PR DESCRIPTION
**Pull request by yogwoggf**
Summary:

- Added all of the scaffold for number matchers.
- Added a corresponding test suite for them.
- `toBeCloseTo`'s implementation is exactly the Jest implementation.

### CODE REVIEW

**Comment by yogwoggf** [comment]
Fuck just realized I need to assert

File: src/matchers/numbers.lua [hunk 1]
```diff
@@ -1,22 +1,128 @@
+local prettyValue = require("src.utils.prettyValue")
+local assertType = require("src.asserts.type")
+
+local INFINITY = math.huge
+
 ---@type lest.Matcher
-local function toBeCloseTo(ctx, received, expected, numDigits) end
+local function toBeCloseTo(ctx, received, expected, numDigits)
+	assertType(received, "number")
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Please do not throw _any_ errors based on the received value. Errors should be thrown only by invalid usage of the matchers in testing instructions, not from the code we're testing.

Also, this does not account for metatables.

File: src/matchers/numbers.lua [hunk 2]
```diff
@@ -1,22 +1,128 @@
+local prettyValue = require("src.utils.prettyValue")
+local assertType = require("src.asserts.type")
+
+local INFINITY = math.huge
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Is this necessary?

File: src/matchers/numbers.lua [hunk 3]
```diff
@@ -1,22 +1,128 @@
...
+	local expectedDiff = math.pow(10, -numDigits) / 2
+	local receivedDiff = math.abs(expected - received)
+	return {
+		pass = receivedDiff < expectedDiff,
+		message = string.format(
+			"Expected %s to%sbe close to %s within %s digits",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected),
+			prettyValue(numDigits)
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Do you need to use `prettyValue` here?

File: src/matchers/numbers.lua [hunk 4]
```diff
@@ -1,22 +1,128 @@
...
+	assertType(received, "number")
+	assertType(expected, "number")
+	assertType(numDigits, "number")
+
+	local expectedDiff = math.pow(10, -numDigits) / 2
+	local receivedDiff = math.abs(expected - received)
+	return {
+		pass = receivedDiff < expectedDiff,
+		message = string.format(
+			"Expected %s to%sbe close to %s within %s digits",
```
Comment (Derpius) [inline src/matchers/numbers.lua]:
```suggestion
			"Expected %s to%sbe close to %s (%s decimal places)",
```

File: src/matchers/numbers.lua [hunk 5]
```diff
@@ -1,22 +1,128 @@
...
+			prettyValue(expected),
+			prettyValue(numDigits)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeGreaterThan(ctx, received, expected) end
+local function toBeGreaterThan(ctx, received, expected)
+	assertType(received, "number")
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Please do not throw any errors based on the received value. Errors should be thrown only by invalid usage of the matchers in testing instructions, not from the code we're testing.

Also, this does not account for metatables.

File: src/matchers/numbers.lua [hunk 6]
```diff
@@ -1,22 +1,128 @@
...
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeGreaterThanOrEqual(ctx, received, expected) end
+local function toBeGreaterThanOrEqual(ctx, received, expected)
+	assertType(received, "number")
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Please do not throw any errors based on the received value. Errors should be thrown only by invalid usage of the matchers in testing instructions, not from the code we're testing.

Also, this does not account for metatables.

File: src/matchers/numbers.lua [hunk 7]
```diff
@@ -1,22 +1,128 @@
...
+	assertType(expected, "number")
+
+	return {
+		pass = received > expected,
+		message = string.format(
+			"Expected %s to%sbe greater than %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
```
Comment (Derpius) [inline src/matchers/numbers.lua:52]: Do you need to use `prettyValue` here?

File: src/matchers/numbers.lua [hunk 8]
```diff
@@ -1,22 +1,128 @@
...
+	assertType(received, "number")
+	assertType(expected, "number")
+
+	return {
+		pass = received >= expected,
+		message = string.format(
+			"Expected %s to%sbe greater than or equal to %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
```
Comment (Derpius) [inline src/matchers/numbers.lua:71]: Do you need to use `prettyValue` here?

File: src/matchers/numbers.lua [hunk 9]
```diff
@@ -1,22 +1,128 @@
...
+	assertType(received, "number")
+	assertType(expected, "number")
+
+	return {
+		pass = received < expected,
+		message = string.format(
+			"Expected %s to%sbe less than %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
```
Comment (Derpius) [inline src/matchers/numbers.lua:90]: Do you need to use `prettyValue` here?

File: src/matchers/numbers.lua [hunk 10]
```diff
@@ -1,22 +1,128 @@
...
+	assertType(received, "number")
+	assertType(expected, "number")
+
+	return {
+		pass = received <= expected,
+		message = string.format(
+			"Expected %s to%sbe less than or equal to %s",
+			prettyValue(received),
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
```
Comment (Derpius) [inline src/matchers/numbers.lua:109]: Do you need to use `prettyValue` here?

File: src/matchers/numbers.lua [hunk 11]
```diff
@@ -1,22 +1,128 @@
...
 ---@type lest.Matcher
-local function toBeNaN(ctx, received) end
+local function toBeNaN(ctx, received)
+	assertType(received, "number")
+
+	return {
+		pass = received ~= received,
+		message = string.format(
+			"Expected %s to%sbe NaN",
+			prettyValue(received),
```
Comment (Derpius) [inline src/matchers/numbers.lua:120]: Do you need to use `prettyValue` here?

File: src/matchers/numbers.lua [hunk 12]
```diff
@@ -1,22 +1,128 @@
...
 ---@type lest.Matcher
-local function toBeInfinity(ctx, received) end
+local function toBeInfinity(ctx, received)
+	assertType(received, "number")
+
+	return {
+		pass = received == INFINITY,
+		message = string.format(
+			"Expected %s to%sbe infinity",
+			prettyValue(received),
```
Comment (Derpius) [inline src/matchers/numbers.lua:132]: Do you need to use `prettyValue` here?

File: src/matchers/numbers.lua [hunk 13]
```diff
@@ -1,22 +1,128 @@
...
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeLessThan(ctx, received, expected) end
+local function toBeLessThan(ctx, received, expected)
+	assertType(received, "number")
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Please do not throw any errors based on the received value. Errors should be thrown only by invalid usage of the matchers in testing instructions, not from the code we're testing.

Also, this does not account for metatables.

File: src/matchers/numbers.lua [hunk 14]
```diff
@@ -1,22 +1,128 @@
...
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeLessThanOrEqual(ctx, received, expected) end
+local function toBeLessThanOrEqual(ctx, received, expected)
+	assertType(received, "number")
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Please do not throw any errors based on the received value. Errors should be thrown only by invalid usage of the matchers in testing instructions, not from the code we're testing.

Also, this does not account for metatables.

File: src/matchers/numbers.lua [hunk 15]
```diff
@@ -1,22 +1,128 @@
...
+			ctx.inverted and " not " or " ",
+			prettyValue(expected)
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeNaN(ctx, received) end
+local function toBeNaN(ctx, received)
+	assertType(received, "number")
```
Comment (Derpius) [inline src/matchers/numbers.lua]: Please do not throw any errors based on the received value. Errors should be thrown only by invalid usage of the matchers in testing instructions, not from the code we're testing.

File: src/matchers/numbers.lua [hunk 16]
```diff
@@ -1,22 +1,128 @@
...
+	}
+end
 
 ---@type lest.Matcher
-local function toBeInfinity(ctx, received) end
+local function toBeInfinity(ctx, received)
+	assertType(received, "number")
+
+	return {
+		pass = received == INFINITY,
```
Comment (Derpius) [inline src/matchers/numbers.lua]: What about negative infinity?

File: src/matchers/numbers.test.lua [hunk 17]
```diff
@@ -0,0 +1,653 @@
...
+				assertMatcher.failed(resultSwapped)
+				assertMatcher.hasMessage(
+					resultSwapped,
+					("Expected %s to be close to %d within 1 digits"):format(
+						tostring(NAN),
+						normalNumber
+					)
+				)
+			end)
+		end)
```
Comment (Derpius) [inline src/matchers/numbers.test.lua:157]: We should also test when both parameters are infinity or `NaN`.

`NaN` should not be close to `NaN`

infinity should be close to infinity, and -infinity should be close to -infinity
infinity should not be close to -infinity
Source: https://github.com/facebook/jest/pull/7444

File: src/matchers/numbers.test.lua [hunk 18]
```diff
@@ -0,0 +1,653 @@
...
+				)
+				assertMatcher.failed(resultSwapped)
+				assertMatcher.hasMessage(
+					resultSwapped,
+					("Expected %s to be greater than %d"):format(
+						tostring(NAN),
+						normalNumber
+					)
+				)
+			end)
```
Comment (Derpius) [inline src/matchers/numbers.test.lua]: For basic operations like greater than or less than, we don't need to assert the behaviour ourselves. The matcher will always behave the same as Lua because it's written in Lua using the built in operators.

File: src/matchers/numbers.test.lua [hunk 19]
```diff
@@ -0,0 +1,653 @@
...
+	describe("toBeInfinity", function()
+		it("should pass when the received value is infinity", function()
+			-- Given
+			local inf = math.huge
+
+			-- When
+			local result = matchers.toBeInfinity(CONTEXT, inf)
+
+			-- Then
+			assertMatcher.passed(result)
```
Comment (Derpius) [inline src/matchers/numbers.test.lua]: Please also test negative infinity here.

Also, there's no need for the arrange step, you can just use `math.huge` directly.
On a more general note, don't try to always have an arrange, act, and assert in every test. For most tests this will be the natural pattern, but there are cases (like this) where skipping one or even two of the sections is perfectly valid.

**Review by Derpius — CHANGES_REQUESTED** [review]
Note, when resolving markups for the matchers, if it adds any functionality then please write tests for it. An example for this would be handling non-number values with numeric operators.

If you want to be fancy about it, you could try doing test driven development. Write the test cases first for the required functionality, then implement the minimum amount of code to pass the tests.

**Comment by yogwoggf** [comment]
Notes:
- Updated `matchers.lua` to include a stack traceback when an assertion fails.
- Fixed all markups and removed the basic Lua operator tests.

**Comment by yogwoggf** [comment]
Oh and of course we discussed that `prettyValue` should stay for the received type because the received value could be infinity or nan which has no numeric representation.

**Comment by yogwoggf** [comment]
bump @Derpius REVIEW

File: src/matchers/numbers.lua [hunk 20]
```diff
@@ -1,22 +1,148 @@
...
+	assertType(numDigits, "number")
+
+	local function createMatcherResult(passed)
+		return {
+			pass = passed,
+			message = string.format(
+				"Expected %s to%sbe close to %s (%d decimal places)",
+				prettyValue(received),
+				ctx.inverted and " not " or " ",
+				prettyValue(expected),
```
Comment (Derpius) [inline src/matchers/numbers.lua:19]: You're asserting that expected is a number, no need to use `prettyValue`

File: src/matchers/numbers.test.lua [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+				local negativeInf = -math.huge
+				local positiveInf = math.huge
+
+				-- When
+				local resultBothInf =
+					matchers.toBeCloseTo(CONTEXT, positiveInf, positiveInf)
+				local resultPosInfNegInf =
+					matchers.toBeCloseTo(CONTEXT, positiveInf, negativeInf)
+				local resultBothNegInf =
+					matchers.toBeCloseTo(CONTEXT, positiveInf, positiveInf)
```
Comment (Derpius) [inline src/matchers/numbers.test.lua]: That is not negative infinity

File: src/matchers/numbers.test.lua (hunk 18 above)
Comment (Derpius) [reply src/matchers/numbers.test.lua]: I meant you don't need to assert how it handles NaN or infinity specifically as that's not affecting the logic.

Instead you deleted all the tests...

File: src/matchers/numbers.test.lua
Comment (Derpius) [inline src/matchers/numbers.test.lua:1]: You're using `INFINITY` in some areas of the tests, and then `math.huge` in others. Do you _really_ need to use `INFINITY` over `math.huge`?

Comment (Derpius) [inline src/matchers/numbers.test.lua:1]: As per my reply to a previous review thread, can you please restore the majority of the tests you deleted, _specifically_ ignoring the ones that tested the behaviour of the built in number operators on infinity and NaN

File: src/asserts/matchers.lua
Comment (Derpius) [inline src/asserts/matchers.lua:1]: This is alright for now, but we should really look into completing https://taservers.atlassian.net/browse/LEST-43 and https://taservers.atlassian.net/browse/LEST-44 instead of doing this here

That being said, there is absolutely no reason to replace the asserts with an if statement and an error call (I have no idea why you did this)

**Review by Derpius — CHANGES_REQUESTED** [review]

**Comment by Derpius** [comment]
FYI there are a bunch of unresolved markups from the _previous_ round of review that you ignored

File: src/matchers/numbers.lua (hunk 20 above)
Comment (yogwoggf) [reply src/matchers/numbers.lua:19]: Fuck off

**Comment by yogwoggf** [comment]
FUCK OFF I HATE THIS STUPID FUCKING TICKET. FUCK YOU. FUCK YOU AND YOUR REVIEWS. 32 FUCKING COMMENTS. ARE YOU SERIOUS. FUCK YOU. I HATE THIOS TICKET. BYE.

**Comment by Derpius** [comment]
@yogwoggf https://exceptionalindividuals.com/candidates/neurodiversity-resources/neurodiversity-quizzes/

**Review by Derpius — APPROVED** [review]
